File: concepts/Refactoring/Java/Java Refactoring Best Practices/OOAbusers/1ConditionalComplexity/BaseProject/src/main/java/com/onlineshop/entities/Order.java
package com.onlineshop.entities;

import com.onlineshop.country.Canada;
import com.onlineshop.country.Country;
import com.onlineshop.items.Item;

import java.util.ArrayList;
import java.util.List;

public class Order {

    private Customer customer;
    private List<Item> items;
    private String voucher;

    public Order(){
        items = new ArrayList<>();
    }

    public int size() {
        return items.size();
    }

    public Order(Customer customer){
        items = new ArrayList<>();
        this.customer = customer;
    }

    public Order(List<Item> items, String voucher){
        this.items = items;
        this.voucher = voucher;
    }

    /**
     * Simple.. if sold only in the US
     */
    public boolean add(Item item){
        if(item.isAgeRestrictedBeverage()){
            if(customer.getAge() < 21) {
                System.out.println("Cannot add age restricted item to order");
                return false;
            }
        }
        return items.add(item);
    }


    public boolean addWithCheck(Item item){
        if(item.isAgeRestrictedBeverage()){
            if(customer.getAge() < getLegalAgeFor(customer)){
                System.out.println("Sorry!");
                return false;
            }
        }

        return items.add(item);
    }


    private int getLegalAgeFor(Customer customer) {
        Country country = customer.getAddress().getCountry();

        if(country instanceof Canada){
            Canada canada = (Canada) country;
            return canada.getLegalDrinkingAge(customer.getAddress().getProvince());
        }


        return country.getMinimumLegalDrinkingAge();
    }



    public List<Item> getItems() {
        return items;
    }

    public void setItems(List<Item> items) {
        this.items = items;
    }

    public String getVoucher() {
        return voucher;
    }

    public void setVoucher(String voucher) {
        this.voucher = voucher;
    }

    public Customer getCustomer() {
        return customer;
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }
}

File: concepts/Refactoring/PHP/PHP Refactoring Best Practices/4Couplers/3ExcessiveExposure/problem/src/entities/Order.php
<?php

namespace app\entities;

use app\country\Canada;

class Order
{
  private $items;
  private $voucher;
  private $customer;

  public function __construct()
  {
    $this->items = [];
  }

  public static function create()
  {
    $instance = new self();
    return $instance;
  }

  public function size()
  {
    return count($this->items);
  }

  public function setItems($items)
  {
    $this->items = $items;
    return $this;
  }

  public function setCustomer($customer)
  {
    $this->customer = $customer;
    return $this;
  }

  public function setVoucher($voucher)
  {
    $this->voucher = $voucher;
    return $this;
  }


  public function getItems()
  {
    return $this->items;
  }

  public function getVoucher()
  {
    return $this->voucher;
  }

  public function getCustomer()
  {
    return $this->customer;
  }


  /**
   * Simple.. if sold only in the US
   */
  public function add(Item $item)
  {
    if ($item->isAgeRestrictedItem()) {
      if ($this->customer->getAge() < 21) {
        // echo "Cannot add age restricted item to order \n";
        return false;
      }
    }
    $this->items[] = $item;
    return true;
  }

  /**
   * More complex version of the same method with more business logic
   * OO is not used
   */
  public function addWithCheck(Item $item)
  {
    if ($item->isAgeRestrictedItem()) {
      $age = $this->customer->getAge();
      if ($age < $this->getLegalAgeFor($this->customer)) {
        echo "Sorry\n";
        return false;
      }
      $items[] = $item;
      return true;
    }
  }

  private function getLegalAgeFor(Customer $customer)
  {
    $country = $customer->getAddress()->getCountry();
    if ($country instanceof Canada) {
      $country->getLegalDrinkingAge($customer->getAddress()->getProvince());
    }
    return $country->getMinimumLegalDrinkingAge();
  }
}
